File: resources/patterns/design/composite-entity-pattern/order_mapper.py
"""Composite Entity Pattern — runnable example with sqlite3.

The OrderMapper treats Order + LineItem + ShippingAddress as one
persistence unit: one save() writes all three tables in a single
transaction, and find_by_id() rehydrates the whole aggregate.

Run: python order_mapper.py  -> prints "Order total: $109.97"
"""
from dataclasses import dataclass, field
from typing import List, Optional
import sqlite3


@dataclass
class LineItem:
    product_id: str
    quantity: int
    unit_price: float

    @property
    def total(self) -> float:
        return self.quantity * self.unit_price


@dataclass
class ShippingAddress:
    street: str
    city: str
    country: str
    postal_code: str


@dataclass
class Order:
    order_id: Optional[str] = None
    customer_id: str = ""
    line_items: List[LineItem] = field(default_factory=list)
    shipping_address: Optional[ShippingAddress] = None

    @property
    def total(self) -> float:
        return sum(item.total for item in self.line_items)


class OrderMapper:
    """Composite entity mapper loading from multiple tables."""

    def __init__(self, conn: sqlite3.Connection):
        self._conn = conn

    def find_by_id(self, order_id: str) -> Optional[Order]:
        row = self._conn.execute(
            "SELECT order_id, customer_id FROM orders WHERE order_id = ?",
            (order_id,),
        ).fetchone()
        if not row:
            return None

        order = Order(order_id=row["order_id"], customer_id=row["customer_id"])

        for item_row in self._conn.execute(
            "SELECT product_id, quantity, unit_price FROM line_items WHERE order_id = ?",
            (order_id,),
        ):
            order.line_items.append(
                LineItem(
                    product_id=item_row["product_id"],
                    quantity=item_row["quantity"],
                    unit_price=item_row["unit_price"],
                )
            )

        addr_row = self._conn.execute(
            "SELECT street, city, country, postal_code FROM shipping_addresses WHERE order_id = ?",
            (order_id,),
        ).fetchone()
        if addr_row:
            order.shipping_address = ShippingAddress(
                street=addr_row["street"],
                city=addr_row["city"],
                country=addr_row["country"],
                postal_code=addr_row["postal_code"],
            )

        return order

    def save(self, order: Order) -> None:
        # One transaction keeps the three tables consistent.
        with self._conn:
            self._conn.execute(
                "INSERT OR REPLACE INTO orders (order_id, customer_id) VALUES (?, ?)",
                (order.order_id, order.customer_id),
            )
            # Delete-then-insert: simplest correct way to handle orphans.
            self._conn.execute(
                "DELETE FROM line_items WHERE order_id = ?", (order.order_id,)
            )
            for item in order.line_items:
                self._conn.execute(
                    "INSERT INTO line_items (order_id, product_id, quantity, unit_price) VALUES (?, ?, ?, ?)",
                    (order.order_id, item.product_id, item.quantity, item.unit_price),
                )
            if order.shipping_address:
                self._conn.execute(
                    """INSERT OR REPLACE INTO shipping_addresses
                       (order_id, street, city, country, postal_code)
                       VALUES (?, ?, ?, ?, ?)""",
                    (
                        order.order_id,
                        order.shipping_address.street,
                        order.shipping_address.city,
                        order.shipping_address.country,
                        order.shipping_address.postal_code,
                    ),
                )


def main() -> None:
    conn = sqlite3.connect(":memory:")
    conn.row_factory = sqlite3.Row
    conn.execute(
        "CREATE TABLE orders (order_id TEXT PRIMARY KEY, customer_id TEXT)"
    )
    conn.execute(
        """CREATE TABLE line_items (
            order_id TEXT, product_id TEXT, quantity INTEGER, unit_price REAL
        )"""
    )
    conn.execute(
        """CREATE TABLE shipping_addresses (
            order_id TEXT PRIMARY KEY, street TEXT, city TEXT,
            country TEXT, postal_code TEXT
        )"""
    )

    mapper = OrderMapper(conn)
    order = Order(
        order_id="ORD-001",
        customer_id="CUST-001",
        line_items=[
            LineItem("PROD-1", 2, 29.99),
            LineItem("PROD-2", 1, 49.99),
        ],
        shipping_address=ShippingAddress(
            "123 Main St", "Springfield", "USA", "62701"
        ),
    )

    mapper.save(order)
    loaded = mapper.find_by_id("ORD-001")
    print(f"Order total: ${loaded.total:.2f}")


if __name__ == "__main__":
    main()

File: resources/patterns/design/composite-entity-pattern/order-mapper.js
// Composite Entity Pattern — Order + LineItem + ShippingAddress persisted
// as one unit across three tables.
//
// Works with an async sqlite wrapper exposing get/all/run/exec
// (e.g. `sqlite` + `sqlite3`, or a thin better-sqlite3 adapter).

class LineItem {
  constructor(productId, quantity, unitPrice) {
    this.productId = productId;
    this.quantity = quantity;
    this.unitPrice = unitPrice;
  }

  get total() {
    return this.quantity * this.unitPrice;
  }
}

class ShippingAddress {
  constructor(street, city, country, postalCode) {
    this.street = street;
    this.city = city;
    this.country = country;
    this.postalCode = postalCode;
  }
}

class Order {
  constructor(orderId, customerId) {
    this.orderId = orderId;
    this.customerId = customerId;
    this.lineItems = [];
    this.shippingAddress = null;
  }

  get total() {
    return this.lineItems.reduce((sum, item) => sum + item.total, 0);
  }
}

class OrderMapper {
  constructor(db) {
    this.db = db;
  }

  async findById(orderId) {
    const row = await this.db.get(
      'SELECT customer_id FROM orders WHERE order_id = ?',
      orderId
    );
    if (!row) return null;

    const order = new Order(orderId, row.customer_id);

    const items = await this.db.all(
      'SELECT product_id, quantity, unit_price FROM line_items WHERE order_id = ?',
      orderId
    );
    for (const item of items) {
      order.lineItems.push(
        new LineItem(item.product_id, item.quantity, item.unit_price)
      );
    }

    const addr = await this.db.get(
      'SELECT street, city, country, postal_code FROM shipping_addresses WHERE order_id = ?',
      orderId
    );
    if (addr) {
      order.shippingAddress = new ShippingAddress(
        addr.street, addr.city, addr.country, addr.postal_code
      );
    }

    return order;
  }

  async save(order) {
    // One transaction keeps the three tables consistent.
    await this.db.exec('BEGIN');
    try {
      await this.db.run(
        'INSERT OR REPLACE INTO orders (order_id, customer_id) VALUES (?, ?)',
        order.orderId, order.customerId
      );
      // Delete-then-insert: simplest correct way to handle orphans.
      await this.db.run('DELETE FROM line_items WHERE order_id = ?', order.orderId);
      for (const item of order.lineItems) {
        await this.db.run(
          'INSERT INTO line_items (order_id, product_id, quantity, unit_price) VALUES (?, ?, ?, ?)',
          order.orderId, item.productId, item.quantity, item.unitPrice
        );
      }
      if (order.shippingAddress) {
        await this.db.run(
          'INSERT OR REPLACE INTO shipping_addresses (order_id, street, city, country, postal_code) VALUES (?, ?, ?, ?, ?)',
          order.orderId, order.shippingAddress.street, order.shippingAddress.city,
          order.shippingAddress.country, order.shippingAddress.postalCode
        );
      }
      await this.db.exec('COMMIT');
    } catch (err) {
      await this.db.exec('ROLLBACK');
      throw err;
    }
  }
}

// --- Usage example (uncomment and point at a real db handle) ---
// const mapper = new OrderMapper(db);
// const order = new Order('ORD-001', 'CUST-001');
// order.lineItems.push(new LineItem('PROD-1', 2, 29.99));
// order.lineItems.push(new LineItem('PROD-2', 1, 49.99));
// order.shippingAddress = new ShippingAddress('123 Main St', 'Springfield', 'USA', '62701');
// await mapper.save(order);
// const loaded = await mapper.findById('ORD-001');
// console.log('Order total: $' + loaded.total.toFixed(2));

module.exports = { LineItem, ShippingAddress, Order, OrderMapper };
